lightning/src/ln/channelmanager.rs

```diff
@@ -3822,14 +3822,47 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			// provide the preimage, so worrying too much about the optimal handling isn't worth
 			// it.
 			let mut claimable_amt_msat = 0;
+			let mut expected_amt_msat = None;
 			let mut valid_mpp = true;
 			for htlc in sources.iter() {
 				if let None = channel_state.as_ref().unwrap().short_to_id.get(&htlc.prev_hop.short_channel_id) {
 					valid_mpp = false;
 					break;
 				}
+				match &htlc.onion_payload {
+					OnionPayload::Spontaneous(_) => {
+						// We don't currently support MPP for spontaneous payments, so just check
+						// that there's one payment here and move on.
+						expected_amt_msat = Some(htlc.value);
+						if sources.len() != 1 {
+							log_error!(self.logger, "Somehow ended up with an MPP spontaneous payment - this should not be reachable!");
+							debug_assert!(false);
+							valid_mpp = false;
+							break;
+						}
+					},
+					OnionPayload::Invoice(hop_data) => {
+						if expected_amt_msat.is_some() && expected_amt_msat != Some(hop_data.total_msat) {
+							log_error!(self.logger, "Somehow ended up with an MPP payment with different total amounts - this should not be reachable!");
+							debug_assert!(false);
+							valid_mpp = false;
+							break;
+						}
+						expected_amt_msat = Some(hop_data.total_msat);
+					},
+				}
+
 				claimable_amt_msat += htlc.value;
 			}
+			if sources.is_empty() || expected_amt_msat.is_none() {
+				log_info!(self.logger, "Attempted to claim an incomplete payment which no longer had any available HTLCs!");
+				return;
+			}
+			if claimable_amt_msat != expected_amt_msat.unwrap() {
+				log_info!(self.logger, "Attempted to claim an incomplete payment, expected {} msat, had {} available to claim.",
```

Comment:
Would the API be more ergonomic if we returned a `Result<(), Error>` from this method, for cases like this?

Reply:
I feel like that sends the wrong impression - an `Ok(())` does not imply that the payment is fully claimed, which is ultimately why the bool was removed from the return value.

Reply:
I guess we expect these errors to be super rare? They do seem useful. Returning an `Option<Error>` is weird though